### DISCUSSION TITLE
MF6 Head Binary File Loaded with Flopy Remains Open and Blocks Folder Deletion

### DISCUSSION THREAD

**Discussion by javgs-bd** [modflowpy/flopy#2291] [Q&A]
As part of a PEST inversion research, I'm running a sequence of MF6 models that are created, executed, postprocessed and then deleted.  I'm loading the head file with:

```` hds_file = os.path.join(model_folder,root_name + '.hds')````

````hds      = flopy.utils.binaryfile.HeadFile(hds_file, model = gwf) ````

Each time, after doing the plots I need, I attempt to remove the model folder with shutil. However, sometimes, the process run into an error like the one below, which I think is related to the file system trying to catch up. I have tried using time.sleep() to wait for some time and then perform the deletion and while this has worked, I wonder why this happens only with the hds file and if there is any way to instruct the code to release the file earlier. 
Is there anything in flopy that should make the head file hanged? or anything that can be done to ensure it is closed?

````
 File c:\modelos\phd\run_inversions.py:8785 in run_inv_sequence
    shutil.rmtree(model_folder)

  File ~\AppData\Local\Programs\Python\Python310\lib\shutil.py:750 in rmtree
    return _rmtree_unsafe(path, onerror)

  File ~\AppData\Local\Programs\Python\Python310\lib\shutil.py:620 in _rmtree_unsafe
    onerror(os.unlink, fullname, sys.exc_info())

  File ~\AppData\Local\Programs\Python\Python310\lib\shutil.py:618 in _rmtree_unsafe
    os.unlink(fullname)

PermissionError: [WinError 32] The process cannot access the file because it is being used by another process: './model\\model.hds'
````

**Comment by rhugman** [comment]
In the past 'del hds' at the end has done the trick for me.

**Reply by javgs-bd** [reply]
Many thanks @rhugman. I'll try that.

**Comment by mwtoews** [comment]
The issue is that `HeadFile()` will open the file, but the open file-handle persists, which may cause PermissionError if attempts to delete/move the file are made while it is open. You would need to explicitly close the file with `hds.close()`.

You could use a "with" statement like this, which will automatically close the file:
```python
with HeadFile(hds_file, model = gwf) as hds:
    hds  # e.g. read, plot, etc.
os.remove(hds_file)
```

**Reply by javgs-bd** [reply]
Thanks @mwtoews. I'm testing this approach now. ````del hds```` worked better, but it still got my inversion process hanged once.
I'll report back.
